sklearn/cluster/_bisect_k_means.py

```diff
@@ -0,0 +1,727 @@
+                samples_mask = leaves_dict[biggest_id]["samples"].copy()
+                samples_mask[picked_samples] = _labels == picked_id
+
+                # Save recently generated leaves
+                leaves_dict[child_id] = {
+                    "samples": samples_mask,
+                    "error_or_size": metrics_values[picked_id],
+                }
+
+            # Split cluster is no longer leaf
+            del leaves_dict[biggest_id]
+
+            last_center_id += 2
+
+            # Pick new 'biggest cluster' to split
+            biggest_id, _ = max(
+                leaves_dict.items(), key=lambda x: x[1]["error_or_size"]
+            )
+
+        # Delete Initial cluster
+        del tree_dict[-1]
+
+        # Sort Leaf centers by their root
+        centers_hierarchical = [0, 1]
+        while len(centers_hierarchical) != self.n_clusters:
+            nested_centers = []
+
+            for center_id in centers_hierarchical:
+                center_ids = (
+                    [center_id]
+                    if tree_dict[center_id]["children"] is None
+                    else tree_dict[center_id]["children"]
+                )
+                nested_centers.extend(center_ids)
+
+            centers_hierarchical = nested_centers
+
+        # Initialize Labels
+        labels = np.full(X.shape[0], -1, dtype=np.intc)
+
+        centers = []
+
+        for i, center_id in enumerate(centers_hierarchical):
+            # Save cluster centers in hierarchical order
+            centers.append(tree_dict[center_id]["center"])
+
+            # Assign labels to proper data points
+            labels[leaves_dict[center_id]["samples"]] = i
+
+        centers = np.asarray(centers)
+
+        # Inner Tree will be later used at 'predict' method:
+        self._inner_tree = tree_dict
+
+        return centers, labels
+
+    def predict(self, X, sample_weight=None):
+        """Predict which cluster each sample in X belongs to by going down
+        the hierarchical tree in searching of closest leaf cluster.
+
+        In the vector quantization literature, `cluster_centers_` is called
+        the code book and each value returned by `predict` is the index of
+        the closest code in the code book.
+
+        Parameters
+        ----------
+        X : {array-like, sparse matrix} of shape (n_samples, n_features)
+            New data to predict.
+
+        sample_weight : array-like of shape (n_samples,), default=None
+            The weights for each observation in X. If None, all observations
+            are assigned equal weight.
+
+        Returns
+        -------
+        labels : ndarray of shape (n_samples,)
+            Index of the cluster each sample belongs to.
+        """
+        check_is_fitted(self)
+
+        X = self._check_test_data(X)
+        x_squared_norms = row_norms(X, squared=True)
+        sample_weight = _check_sample_weight(sample_weight, X, dtype=X.dtype)
```

Comment:
It is not clear to me how sample weights are used during prediction. Isn't each sample independent during prediction?

Reply:
`lloyd_iter_chunked_sparse` and `lloyd_iter_chuncked_dense` which comes from `KMeans` and are used for assigning labels requires this param.

Looking into cython code seems that `sample_weight` as it should  - is only used for updating centers.
I've kept it for consistancy with `KMeans.predict` but as optimization I see two options:
- create cython function like `lloyd_iter_chunked_sparse` and `lloyd_iter_chuncked_dense` but without `sample_weight` and `update_centers`
- Only generate `sample_weight` as ones or even zeros, without calling `_check_sample_weight`

What do you think @TomDLT ?